### DISCUSSION TITLE
Move depthcharge related stuff out of imagebuilder

### DISCUSSION THREAD

**Discussion by s1gnate-sync** [velvet-os/imagebuilder#242] [General]
Hey hexdump0815,

I recently discovered interesting project https://github.com/FyraLabs/submarine which makes depthcharge kpart act as bootloader so you can load your distro without additional depthcharge related tweaks. It seems like a good direction to go as it makes bootstrapping new distro much easier, you just need to write it once.

Do you have any thoughts on it?

My opinion - I like the project and overall idea with kexec, but dislike how u-root works and looks... I really struggle to see any advan tage over classic approach. So just for fun I tried to make minimalistic kexec bootloader based on alpine linux. 

https://github.com/s1gnate-sync/sideboot/tree/main

**Comment by thenameisluk** [comment]
@s1gnate-sync depthcharge is basically booting kernel directly, submarine is more of grub approach

while project seams nice at first it would make velvet for chromebooks completely dependent on submarine instead of using the _somewhat_ officialy supported boot system

it wouldn't still work with kernel distributed by all distributions since these don't apple patches some chromebooks require, we would still need to ship out own version, for example [trogdor](https://github.com/hexdump0815/linux-mainline-and-mali-generic-stable-kernel/blob/master/readme.cbq) or [kukui](https://github.com/hexdump0815/linux-mainline-mediatek-mt81xx-kernel/blob/master/readme.mt8)

unless you are planning to dual boot you don't really need this

the depthcharge stuff only requires you to specify root partition and it boots any distro you want, ppl for example made [arch boot](https://github.com/panbed/imagebuilder)

also i don't really see kexec as stable thing (imo) and using it instead of booting directly might introduce side effects or cause issues with hardware initialisation (afaik)

it would be more beneficial for "letting ppl use different distros" cause to make depthcharge stuff we have more generic, but still even then someone would have to maintain these distros since the chromebooks require some blobs to function right

i've been working on [tools](https://github.com/thenameisluk/velvettools) that let you build bootable image and flash or test it automatically, it's simpler to do and doesn't really introduce any more abstraction (more of an automation) and it's been working really well for me with these images, it can also be automatically update using [repo](https://github.com/thenameisluk/velvet-repo) i've made that can also distribute required patched linux

why would you need additional bootloader instead of booting directly? what is it trying to solve?

**Reply by s1gnate-sync** [reply]
Thanks for the reply... I'm experimenting to see if it's possible to run distro like on UEFI with little to none modification. I wonder why so much x86_64 devices work with default kernel while arm devices not? 

To make it sound simple I want to have be able boot from iso, live cd and so on

**Reply by thenameisluk** [reply]
> Thanks for the reply... I'm experimenting to see if it's possible to run distro like on UEFI with little to none modification. I wonder why so much x86_64 devices work with default kernel while arm devices not? 

it's not really dependent on devices
it's just the way linux is developed
each device needs different set of drivers and stuff in the kelner
chromebooks run chromeos and its kelner has all required patches the device needs
in order for device to work with normal linux the patches must get upstreamed (added to the main repo)

there are arm64 that can boot from iso for instance lenovo x13s but these use more traditional boot system while chromebooks use depthcharge

different devices use different boot systems like uefi or mbr. you can't just put iso onto usb and have every device boot from it, you need an image that your bios can boot

> To make it sound simple I want to have be able boot from iso, live cd

you can flash these images onto usb and just boot them since these are made for chromebooks, you can even install encrypted variant on it if you want or put together a distro of your choosing based on them

just modify the scripts in these repo to bootstrap distro of your choosing in there, unless you want to distro hop :3 it shouldn't be too hard

**Reply by s1gnate-sync** [reply]
Haha thanks! Regarding different devices need different drivers... I thought it's all sorted using separate firmware files and kernel modules... By default include everything in the kernel, but as a module. It won't hurt, right?

**Reply by thenameisluk** [reply]
@s1gnate-sync you can't use module to patch kernel
for instance
```
> # fix for at least the panel on some wormdingler devices (maybe others too)
> patch -p1 < /compile/doc/stable/misc.cbq/patches/fix-boe-tv101wum-nl6-panel.patch
```

if u can live with half working system you can use default version but if u want more stuff to work u need these patches

**Comment by hexdump0815** [comment]
for the near and medium future i would stay with depthcharge for sure - i'm aware of submarine and anyone interested is free to play around with it based on these images here (and share the results/experiences here), but i prefer to see it more stabilizing before thinking about switching ... i also did experiments with kexec booting long time ago (see https://github.com/hexdump0815/petitboot ) and dropped the idea as it was not working very consistent across different systems - for chromebooks we can be sure that depthcharge will work for all of them (we use it for the 64bit chromebooks) and for all 32bit chromebooks which exist we have a u-boot we can chainload, so we are fine on those as well (as depthcharge is a bit problematic on those due to various size limitations)

ps: and one note regarding x86 vs arm64 - x86 uses acpi for describing the hardware and x86 linux has very good support for that - on arm64 acpi is useless for all more complex drivers (as many of them are not even properly described in the acpi tables anymore) ... for that reason arm64 uses devicetrees to describe the hardware in linux, so a completely different approach
